commons/src/main/java/org/infinispan/commons/api/functional/Traversable.java

```diff
@@ -0,0 +1,215 @@
+package org.infinispan.commons.api.functional;
+
+import org.infinispan.commons.api.functional.Closeables.CloseableSpliterator;
+import org.infinispan.commons.util.CloseableIterator;
+
+import java.util.Comparator;
+import java.util.Optional;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.BinaryOperator;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+/**
+ * Unsorted traversable stream. Design is inspired on {@link java.util.stream.Stream}.
+ * Some functions that would be typically found in {@link java.util.stream.Stream}
+ * are not included here:
+ *
+ * <ul>
+ *    <li>{@link java.util.stream.Stream#distinct()}: This operation would be
+ *    extremely costly in a distributed environment as all data up to that
+ *    point of the traversable stream must be read by a single node.
+ *    Due to the memory and performance constraints this method wouldn't be supported.
+ *    </li>
+ *    <li>{@link java.util.stream.Stream#sorted()}: Providing a sorted
+ *    traversable stream in a distributed environment is an expensive operation.
+ *    </li>
+ *    <li>{@link java.util.stream.Stream#limit(long)}: In a distributed
+ *    environment this would require sequential requests for each node to
+ *    verify that we haven't reached the limit yet. This could be optimized
+ *    depending on the terminator. This operation makes most sense when
+ *    combined with sorted streams.
+ *    </li>
+ *    <li>{@link java.util.stream.Stream#sorted()}: Operation that makes
+ *    most sense when combined with sorting.
+ *    </li>
+ *    <li>{@link java.util.stream.Stream#forEachOrdered(Consumer)}:
+ *    Sorting related.
+ *    </li>
+ *    <li>{@link java.util.stream.Stream#toArray()}: Creating an array
+ *    requires size to be computed first, and then populate it, making it a
+ *    potentially expensive operation in a distributed operation. Also,
+ *    there are other ways to achieve this, such as mapping to a variable
+ *    length collection and then narrowing it down to an array?
+ *    </li>
+ *    <li>{@link java.util.stream.Stream#collect(java.util.stream.Collector)}:
+ *    This method is designed to work with {@link java.util.stream.Collectors},
+ *    whose static methods are used as helpers functions, but these are not
+ *    Serializable, and we'd need to investigate whether we can provide
+ *    marshallable versions. Also, {@link java.util.stream.Collectors} brings
+ *    in a lot of baggage. Instead, {@link #collect(Supplier, BiConsumer, BiConsumer)}
+ *    should be used.
+ *    </li>
+ *    <li>{@link java.util.stream.Stream#findFirst()}: The only difference
+ *    with {@link #findAny()} is ordering, so since we do not want to enter
+ *    the ordering business, we leave it out.</li>
```

Comment:
Sorting can be quite expensive in any case.  The problem is that you have to bring all of the results from each node to the local node first before performing the terminal operation.  If you see [1] I can perform all of the intermediate operations up to the sort remotely and then use an iterator to pull those results down and finally push those results into a new stream where any additional intermediate operations after sorted were applied and finally the terminal operation.  Filter can help by reducing the number of elements returned, but the core issue is still present.  Also reduction is a non ordered terminal operation and as such I ignore sorting for DistributedStreams when that terminal operator is used.

[1] https://github.com/infinispan/infinispan/pull/3569/files#diff-257b456b635d0d1e652ed3adc904fd11R717